### PR TITLE
Add a hook that runs periodically

### PR DESCRIPTION
**Pull request by dcbaker**
This adds the much sough after functionality to run arbitrary commands on a timer. This is acheived using twisted's event loop (the reactor), and a `LoopingCall`. Basic error handling is provided as well (errors are sent to the log).

### CODE REVIEW

File: docs/source/configuration/hooks.rst [hunk 1]
```diff
@@ -194,3 +194,12 @@ Apart from command pre- and posthooks, the following hooks will be interpreted:
     :type suffix: str
     :returns: tuple of `prefix` and `suffix`
     :rtype: (str, str)
+
+.. py:function:: loop_hook(ui=None)
+
+    Run on a period controlled by settings:loop_period.
```
Comment (pazz) [inline docs/source/configuration/hooks.rst]: you can use 
```
:ref:`loop_hook <loop-hook>`
```
to refer to your new config option here.

File: docs/source/configuration/hooks.rst [hunk 2]
```diff
@@ -194,3 +194,12 @@ Apart from command pre- and posthooks, the following hooks will be interpreted:
...
+
+    All errors are logged by default.
```
Comment (pazz) [inline docs/source/configuration/hooks.rst]: I don't think this is necessary to say. kind of obvious no?

File: alot/defaults/alot.rc.spec [hunk 3]
```diff
@@ -248,6 +248,9 @@ reply_account_header_priority = force_list(default=list(From,To,Cc,Envelope-To,X
 #           history file might get *very* long.
 history_size = integer(default=50)
 
+# The number of seconds to wait between calls to the loop_hook
+loop_period = integer(default=300)
```
Comment (pazz) [inline alot/defaults/alot.rc.spec]: I'm not a fan of the name: How about "periodic_hook_frequency"?
although technically this describes the period, not a frequency.
By the way. Are you sure that this is in seconds, not milliseconds?

File: alot/settings/manager.py [hunk 4]
```diff
@@ -311,8 +311,7 @@ def colourpick(triple):
     def get_hook(self, key):
         """return hook (`callable`) identified by `key`"""
         if self.hooks:
-            if key in self.hooks.__dict__:
-                return self.hooks.__dict__[key]
+            return getattr(self.hooks, key, None)
```
Comment (lucc) [inline alot/settings/manager.py]: Isn't None the default?

Comment (dcbaker) [inline alot/settings/manager.py]: No, with `getattr` if you don't provide a fallback then an `AttributeError` is raised.
https://docs.python.org/2/library/functions.html#getattr

File: alot/defaults/alot.rc.spec (hunk 3 above)
Comment (dcbaker) [inline alot/defaults/alot.rc.spec]: what about "loop_hook_frequency" or "loop_hook_period"?

Yes, `LoopingCall`'s interval argument is defined in seconds.

**Comment by pazz** [comment]
Quoting Dylan Baker (2017-01-26 18:36:20)
>    what about "loop_hook_frequency" or "loop_hook_period"?

I was trying to avoid "loop" because it sounds like "main loop" to me..
Ho about "periodic_hook_sleeps" ?
In any case, it's not the end of the world for me if we go with "loop"
:)

More important is the possibility to switch this feature off. Maybe by
setting this config setting to 0?

>    Yes, LoopingCall's interval argument is defined in seconds.

OK thanks!

**Comment by dcbaker** [comment]
> I was trying to avoid "loop" because it sounds like "main loop" to me..
Ah, okay. In that case I like your original idea better.

> More important is the possibility to switch this feature off. Maybe by setting this config setting to 0?
If you don't define a loop_hook then this won't get enabled. Is that enough, or would you like an additional switch?

**Comment by pazz** [comment]
Quoting Dylan Baker (2017-01-26 18:51:22)
>      I was trying to avoid "loop" because it sounds like "main loop" to me..
>      Ah, okay. In that case I like your original idea better.

cool, your call.

>      More important is the possibility to switch this feature off. Maybe by
>      setting this config setting to 0?
>      If you don't define a loop_hook then this won't get enabled. Is that
>      enough, or would you like an additional switch?

Ah OK, so this defaults to `None`? I think, sinse a period of $0$ makes
little sense anyway, we could additionally interpret 0 to disable the
feature.

**Comment by dcbaker** [comment]
> Up for discussion: Do you think that people will want to run several different periodic hooks or would one suffice?

One of the ideas I've had is to replace the hook module with a hook package, which would allow people to implement multiple functions called at the same time. The advantage of this would be that if you need to have cleanups you can implement a single module with the right hooks for a feature.

> Ah OK, so this defaults to `None`? I think, sinse a period of $0$ makes
little sense anyway, we could additionally interpret 0 to disable the
feature.

Right, it uses `settings.get_hook`, and then does 'if hook', so if there was no hook defined you get no loop. I don't know how useful it is to have a settings that says "I defined a loop_hook, but please don't run it."

File: alot/settings/manager.py (hunk 4 above)
Comment (lucc) [inline alot/settings/manager.py]: ok, sorry mixed it up with stuff like dict.get()

Comment (dcbaker) [inline alot/settings/manager.py]: Yup. I had to look to see. It's strange that getattr and dict.get behave differently given that getattr is effecitvely a wrapper around dict.get.

**Pull request merged**